proposals/000-decorator-api.md

```diff
@@ -0,0 +1,322 @@
+    // Present a highlight pop-up for `decoration`...
+}
+```
+
+### Registering Decoration Styles
+
+Reading apps are welcome to register new decoration styles for custom use cases. The API depends on each Navigator, but here's an example using an HTML Navigator (e.g. for EPUB) to implement a sidemark:
+
+```swift
+// Declare a new style with associated configuration values.
+class Decoration.Style.Sidemark {
+    let tint: Color?
+}
+
+// Create a new HTML decoration template which will be used with our new style.
+let sidemarkTemplate = HTMLDecorationTemplate(
+    layout: .bounds,
+    width: .page,
+    element: { decoration in
+        let style = decoration.style as? Decoration.Style.Sidemark
+        let tint = style?.tint ?? Color.red
+        """
+        <div><div class="sidemark" style="--tint: \(tint.cssColor)"/></div>
+        """,
+    },
+    stylesheet: """
+        .sidemark {
+            float: left;
+            width: 5px;
+            height: 100%;
+            background-color: var(--tint);
+            margin-left: 20px;
+            border-radius: 3px;
+        }
+        [dir=rtl] .sidemark {
+            float: right;
+            margin-left: 0px;
+            margin-right: 20px;
+        }
+        """
+)
+
+// Associate the style with our template when creating the HTML navigator.
+var config = HTMLNavigator.Configuration()
+config.decorationStyles[Decoration.Style.Sidemark] = sidemarkTemplate
+let navigator = HTMLNavigator(config)
+```
+
+![](assets/000-sidemark.png)
+
+### Checking Whether a Navigator Supports a Decoration Style
+
+You should check whether the Navigator supports drawing the decoration styles required by a particular feature before enabling it. For example, underlining an audiobook does not make sense, so an Audiobook Navigator would not support the `underline` decoration style.
+
+```swift
+navigator.supportsDecorationStyle(Decoration.Style.Underline)
+```
+
+### Backward Compatibility and Migration
+
+#### Kotlin
+
+The Readium Kotlin toolkit currently ships with a highlighting API. It will be deprecated in favor of the new Decorator API. The old APIs will internally use the Decorator implementation, so it will not be a breaking change.
+
+#### Swift
+
+The Readium Swift toolkit did not yet have decoration capabilities, so there's no impact on existing implementations.
+
+
+## Reference Guide
+
+### `DecorableNavigator` Interface
+
+A navigator able to render arbitrary decorations over a publication.
+
+#### Methods
+
+* `apply(decorations: [Decoration], group: String)`
+    * Declares the current state of the decorations in the given decoration `group`.
+    * Name each decoration group as you see fit. A good practice is to use the name of the feature requiring decorations, e.g. `annotation`, `search`, `tts`, etc.
+    * The Navigator will decide when to actually render each decoration efficiently. Your only responsibility is to submit the updated list of decorations when there are changes.
```

Comment:
I guess this is platform-specific. I believe some kind of observable list should be a good fit at least on Kotlin.

Reply:
I envisioned this exactly like [`ListAdapter.submitList()`](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter#submitList(java.util.List%3CT%3E)). I'm not sure how using an observable list would make this nicer to use or simpler.

If you have a `Flow` or `LiveData`, you can use the common pattern that is used with `ListAdapter` as well:

```kotlin
highlights.observe(viewLifecycleOwner) {
    navigator.apply(highlights, group = "highlights")
}
```

Reply:
Yeah, that's right. It's better to remain neutral about the kind of observable apps would use.